### PR TITLE
[CSBindings] Detect situations when type variable bindings could be i…

### PR DESCRIPTION
**Pull request by xedin**
…ncomplete

When `bind param` constraint is associated with a given type
variable a set of bindings collected for it is potentially
incomplete, because binding gathering doesn't look through related
type variables. Which means that such type variable has to be
de-prioritized until `bind param` constraint is resolved
or there is just nothing else to try, otherwise there is a
risk that solver would skip some of the valid solutions.

Resolves: rdar://problem/45659733

### CODE REVIEW

**Comment by xedin** [comment]
In the future we'd be better off merging `fully bound` and other flags with this "incomplete" flag.

**Comment by xedin** [comment]
@swift-ci please test

**Comment by xedin** [comment]
@swift-ci please test source compatibility

File: test/Constraints/generics.swift
Comment (slavapestov) [inline test/Constraints/generics.swift]: Maybe this belongs more in test/Constraints/closures.swift, but I'm not too picky.

**Review by slavapestov — APPROVED** [review]

File: test/Constraints/generics.swift
Comment (xedin) [reply test/Constraints/generics.swift]: You are right, I think closures might be better place for this!

**Comment by xedin** [comment]
@swift-ci please test

**Comment by xedin** [comment]
@swift-ci please test source compatibility

**Comment by xedin** [comment]
@swift-ci please smoke test compiler performance

**Comment by xedin** [comment]
@swift-ci please smoke test compiler performance

**Review by DougGregor — APPROVED** [review]

File: lib/Sema/ConstraintSystem.h
Comment (xedin) [reply lib/Sema/ConstraintSystem.h]: I actually tried to make it so FullyBound is set when there is a bind param constraint but it breaks ordering even worse (tried a bunch of different combinations), I’m hoping to re-visit fully bound soon. Trying to unblock some code with this PR...

Comment (xedin) [reply lib/Sema/ConstraintSystem.h]: Maybe the best possible answer here is to re-work bind param itself, so we don’t get into situations like that, or make some of the choices retry-able

**Comment by xedin** [comment]
@swift-ci please smoke test

**Comment by xedin** [comment]
@swift-ci please smoke test

**Comment by xedin** [comment]
I'll run smoke test one more time before merging, just in case...

**Comment by xedin** [comment]
@swift-ci please smoke test Linux platform

**Comment by xedin** [comment]
@swift-ci please smoke test Linux platform

**Comment by xedin** [comment]
@swift-ci please smoke test Linux platform

**Pull request merged**
